packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart

```diff
@@ -89,74 +40,79 @@ class FlutterCodeAsset {
 
   final CodeAsset codeAsset;
   final Target target;
-
-  @override
-  String toString() =>
-      'FlutterCodeAsset(codeAsset: ${codeAsset.id} ${codeAsset.file}, target: $target)';
 }
 
-/// Invokes the build of all transitive Dart packages and prepares code assets
+enum SupportedAssetTypes { codeAssets, dataAssets }
+
+/// Invokes the build of all transitive Dart package hooks and prepares assets
 /// to be included in the native build.
-Future<DartBuildResult> runFlutterSpecificDartBuild({
+Future<DartHooksResult> runFlutterSpecificHooks({
   required Map<String, String> environmentDefines,
   required FlutterNativeAssetsBuildRunner buildRunner,
   required TargetPlatform targetPlatform,
   required Uri projectUri,
   required FileSystem fileSystem,
 }) async {
-  final OS targetOS = getNativeOSFromTargetPlatform(targetPlatform);
-  final Uri buildUri = nativeAssetsBuildUri(projectUri, targetOS);
+  final Uri buildUri = nativeAssetsBuildUri(projectUri, targetPlatform.osName);
   final Directory buildDir = fileSystem.directory(buildUri);
-
-  final bool flutterTester = targetPlatform == TargetPlatform.tester;
-
   if (!await buildDir.exists()) {
     // Ensure the folder exists so the native build system can copy it even
     // if there's no native assets.
     await buildDir.create(recursive: true);
   }
 
-  if (!await _nativeBuildRequired(buildRunner)) {
-    return const DartBuildResult.empty();
+  if (!await _hookRunRequired(buildRunner)) {
+    return DartHooksResult.empty();
   }
 
-  final BuildMode buildMode = _getBuildMode(environmentDefines, flutterTester);
-  final List<Architecture> architectures =
-      flutterTester
-          ? <Architecture>[Architecture.current]
-          : _architecturesForOS(targetPlatform, targetOS, environmentDefines);
-  final DartBuildResult result =
-      architectures.isEmpty
-          ? const DartBuildResult.empty()
-          : await _runDartBuild(
-            environmentDefines: environmentDefines,
-            buildRunner: buildRunner,
-            architectures: architectures,
-            projectUri: projectUri,
-            linkingEnabled: _nativeAssetsLinkingEnabled(buildMode),
-            fileSystem: fileSystem,
-            targetOS: targetOS,
-          );
-  return result;
+  final List<SupportedAssetTypes> supportedAssetTypes = <SupportedAssetTypes>[
+    if (featureFlags.isNativeAssetsEnabled) SupportedAssetTypes.codeAssets,
+    if (featureFlags.isDartDataAssetsEnabled) SupportedAssetTypes.dataAssets,
+  ];
+  final List<AssetBuildTarget> targets = AssetBuildTarget.targetsFor(
+    targetPlatform,
+    environmentDefines,
+    fileSystem,
+    supportedAssetTypes,
+  );
+
+  // This is ugly, but sadly necessary as fetching the cCompilerConfig is async,
+  // while using it in native_assets_builder is not.
+  for (final CodeAssetTarget target in targets.whereType<CodeAssetTarget>()) {
```

Comment:
Does this need to be part of the build runner interface if we're basically just calling `target.setCCompilerConfig()`?

Also, if we expect for there to be more than a single target here, we should instead do this to avoid awaiting on consecutive futures that aren't dependent on each other:

```dart
await Future.wait<void>(<Future<void>[
  for (final CodeAssetTarget target in targets.whereType<CodeAssetTarget>())
    buildRunner.setCCompilerConfig(target)
]);
```

Reply:
I think `Future.wait` swallows stack traces and regurgitates them with a different exception type, which is why I used this instead. Also, we have at most two targets, so the overhead should be fine, and this way the code is simpler.

Reply:
Regarding the interface, I am open to other solutions - this is ugly after all.